### PR TITLE
Process credentials

### PR DESCRIPTION
**Pull request by tahoward**
*Description of changes:*

Add a new config section to implement Aws::ProcessCredentials.

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

**Comment by riywo** [comment]
Hi,

Thank you for this PR. We are prioritizing now.

It is not mandatory, but do you have a chance to add unit tests like this? https://github.com/awslabs/aws-fluent-plugin-kinesis/blob/master/test/kinesis_helper/test_client.rb#L48-L62

**Comment by riywo** [comment]
Also, if possible, could you squash your commits to a single (or a few) commit(s)?

**Comment by riywo** [comment]
For maintainers: we should point this (or equivalent) link in README to describe danger of the feature. https://docs.aws.amazon.com/cli/latest/topic/config-vars.html#sourcing-credentials-from-external-processes

**Comment by tahoward** [comment]
@riywo Thank you for looking into this! I have implemented suggested changes.

File: lib/fluent/plugin/kinesis_helper/client.rb [hunk 1]
```diff
@@ -63,6 +63,10 @@ module ClientParams
             desc "Profile name. Default to 'default' or ENV['AWS_PROFILE']"
             config_param :profile_name, :string, default: nil
           end
+          config_section :process_credentials, multi: false do
+            desc "External process to execute."
+            config_param :process, :string, default: nil
```
Comment (riywo) [inline lib/fluent/plugin/kinesis_helper/client.rb]: This config seems mandatory to use `process_credentials`. Is it ok to strip `default: nil`?

File: lib/fluent/plugin/kinesis_helper/client.rb [hunk 2]
```diff
@@ -146,6 +150,10 @@ def setup_credentials
             credentials_options[:path] = c.path if c.path
             credentials_options[:profile_name] = c.profile_name if c.profile_name
             options[:credentials] = Aws::SharedCredentials.new(credentials_options)
+          when @process_credentials
+            c = @process_credentials
+            process = c.process if c.process
```
Comment (riywo) [inline lib/fluent/plugin/kinesis_helper/client.rb]: Same above. What happens if `process` is `nil`?

File: README.md [hunk 3]
```diff
@@ -241,6 +241,27 @@ Path to the shared file. Defaults to "#{Dir.home}/.aws/credentials".
...
+        process CMD
+      </process_credentials>
+    </match>
+
+See also:
+
+*   [Aws::ProcessCredentials](https://docs.aws.amazon.com/sdk-for-ruby/v3/api/Aws/ProcessCredentials.html)
+*   [Sourcing Credentials From External Processes](https://docs.aws.amazon.com/cli/latest/topic/config-vars.html#sourcing-credentials-from-external-processes)
+
+**process**
```
Comment (simukappu) [inline README.md]: Agree with @riywo. I think `process` param should be `required`.

Comment (simukappu) [reply README.md]: Same as `role_arn` param in `assume_role_credentials`.
https://github.com/awslabs/aws-fluent-plugin-kinesis/pull/178/files/0e82e7fe74ca7241f25bad9ff5e843a173827bc5#diff-04c6e90faac2675aa89e2176d2eec7d8R152

Comment (tahoward) [reply README.md]: Woops, yes that should be required. I'll update this.

File: test/kinesis_helper/test_client.rb [hunk 4]
```diff
@@ -89,9 +125,9 @@ def setup_instance_profile(expiration = Time.now.utc + 3600)
     JSON
```
Comment (simukappu) [inline test/kinesis_helper/test_client.rb]: Please do not change this setup_instance_profile function. It is unrelated.

File: test/kinesis_helper/test_client.rb [hunk 5]
```diff
@@ -46,21 +72,31 @@ def teardown
...
     setup_instance_profile(Time.now.utc + 299)
-    credentials = @object.client.config.credentials
+    credentials = MockClientInstanceProfile.new.client.config.credentials
     assert_equal 'akid-2',          credentials.credentials.access_key_id
     assert_equal 'secret-2',        credentials.credentials.secret_access_key
     assert_equal 'session-token-2', credentials.credentials.session_token
   end
 
+  def test_process_credentials
+    omit_if(Gem::Version.new(Aws::CORE_GEM_VERSION) < Gem::Version.new('3.24.0'))
```
Comment (riywo) [inline test/kinesis_helper/test_client.rb]: We need the same validation on `client.rb` as well. If the core SDK version is too old, it should throw configuration error to the customer.

Comment (riywo) [inline test/kinesis_helper/test_client.rb]: We need the same validation on `client.rb` as well. If the SDK version is too old, it should throw configuration error to the customer.

**Comment by tahoward** [comment]
Added suggested ConfigError for aws-sdk-core < 3.24.0

**Comment by riywo** [comment]
We've released v3.1.0 today. Thank you for your all contribution! https://rubygems.org/gems/fluent-plugin-kinesis/versions/3.1.0

**Pull request merged**